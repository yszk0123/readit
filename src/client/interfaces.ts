import { Action as ReduxAction } from 'redux';

// Type

export type ID = string;

export type DateOnly = string;

export enum ReviewStatus {
  PLAN_TO_BUY = 'PLAN_TO_BUY',
  UNREAD = 'UNREAD',
  READING = 'READING',
  READ = 'READ',
  STOP = 'STOP',
}

export type ReviewRating = 0 | 1 | 2 | 3 | 4 | 5;

// Entity

export interface User {
  id: ID;
  nickname: string;
  email: string | null;
}

export interface Book {
  id: ID;
  title: string;
  subtitle: string | null;
  author: string;
  thumbnailLink: string;
  description: string;

  category: string;
  pageCount: number;
  publisher: string;
  publishedAt: DateOnly;
  isbn: string | null;
  meta: {
    source: {
      type: string;
      link: string;
      searchText: string;
    };
  };
}

export interface Review {
  id: ID;
  status: ReviewStatus;
  rating: ReviewRating;
}

export interface ReadingLog {
  id: ID;
  book: Book;
  review: Review;
  removed?: boolean;
}

// Action

export interface Action extends ReduxAction {
  payload: any;
}

export enum ActionTypes {
  HELLO = 'HELLO',
  CHANGE_NAME = 'CHANGE_NAME',
  FETCH_SUCCESS = 'FETCH_SUCCESS',
  FETCH_FAILURE = 'FETCH_FAILURE',
  FETCH_READING_LOGS = 'FETCH_READING_LOGS',
  CREATE_READING_LOG = 'CREATE_READING_LOG',
  UPDATE_REVIEW = 'UPDATE_REVIEW',
  SEARCH_BOOK_BY_TITLE = 'SEARCH_BOOK_BY_TITLE',
  REMOVE_READING_LOG = 'REMOVE_READING_LOG',
  REMOVE_READING_LOG_SUCCESS = 'REMOVE_READING_LOG_SUCCESS',
  REMOVE_ENTITY = 'REMOVE_ENTITY',
}

// State

export interface State {
  entities: {
    users: Record<ID, User>;
    books: Record<ID, Book>;
    reviews: Record<ID, Review>;
    readingLogs: Record<ID, ReadingLog>;
  };
  ui: {
    currentUser: ID | null;
    readingLogs: ID[];
    titleInput: string;
  };
}
