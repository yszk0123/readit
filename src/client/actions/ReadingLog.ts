import { ID, Action, ActionTypes } from '../interfaces';

export type ReadingLogAction = FetchAll | Create | Remove;

interface FetchAllPayload {
  limit?: number;
}

export interface FetchAll extends Action {
  type: ActionTypes.FETCH_READING_LOGS;
  payload: FetchAllPayload;
}

export function fetchAll(payload: FetchAllPayload): FetchAll {
  return { type: ActionTypes.FETCH_READING_LOGS, payload };
}

interface CreatePayload {
  title: string;
}

export interface Create {
  type: ActionTypes.CREATE_READING_LOG;
  payload: CreatePayload;
}

export function create(payload: CreatePayload): Create {
  return { type: ActionTypes.CREATE_READING_LOG, payload };
}

export interface Remove {
  type: ActionTypes.REMOVE_READING_LOG;
  payload: { readingLogId: ID };
}

export function remove(readingLogId: ID): Remove {
  return { type: ActionTypes.REMOVE_READING_LOG, payload: { readingLogId } };
}

export interface RemoveSuccess {
  type: ActionTypes.REMOVE_READING_LOG_SUCCESS;
  payload: { readingLogId: ID };
}

export function removeSuccess(readingLogId: ID): RemoveSuccess {
  return {
    type: ActionTypes.REMOVE_READING_LOG_SUCCESS,
    payload: { readingLogId },
  };
}
