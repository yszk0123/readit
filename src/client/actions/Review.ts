import { ID, ActionTypes, ReviewStatus, ReviewRating } from '../interfaces';

export type ReviewAction = Update;

type UpdatePayload = Partial<{
  reviewId: ID;
  status: ReviewStatus;
  rating: ReviewRating;
}>;

export interface Update {
  type: ActionTypes.UPDATE_REVIEW;
  payload: UpdatePayload;
}

export function update(payload: UpdatePayload): Update {
  return { type: ActionTypes.UPDATE_REVIEW, payload };
}
