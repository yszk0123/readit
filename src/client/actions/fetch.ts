import { ActionTypes, State } from '../interfaces';

export type FetchAction = FetchSuccessAction | FetchFailureAction;

type FetchableState = Partial<{
  entities: Partial<State['entities']>;
  ui: Partial<State['ui']>;
}>;

export interface FetchSuccessAction {
  type: ActionTypes.FETCH_SUCCESS;
  payload: FetchableState;
}

export function fetchSuccess(payload: FetchableState): FetchSuccessAction {
  return { type: ActionTypes.FETCH_SUCCESS, payload };
}

export interface FetchFailureAction {
  type: ActionTypes.FETCH_FAILURE;
  payload: Error;
}

export function fetchFailure(payload: Error): FetchFailureAction {
  return { type: ActionTypes.FETCH_FAILURE, payload };
}
