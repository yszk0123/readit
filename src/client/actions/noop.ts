import { ActionTypes } from '../interfaces';

export interface NoopAction {
  type: ActionTypes.NOOP;
  payload: {};
}

export function noop(): NoopAction {
  return { type: ActionTypes.NOOP, payload: {} };
}
