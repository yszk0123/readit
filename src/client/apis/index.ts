import axios from 'axios';

export async function fetch(uri: string) {
  const { data } = await axios.get(uri);

  return data;
}
