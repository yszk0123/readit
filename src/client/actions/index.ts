import * as ReadingLog from './ReadingLog';
import * as Review from './Review';
import { SearchBookByTitleAction } from './searchBookByTitle';

export type Action =
  | ReadingLog.ReadingLogAction
  | Review.ReviewAction
  | SearchBookByTitleAction;

export { ReadingLog, Review };

export * from './searchBookByTitle';
