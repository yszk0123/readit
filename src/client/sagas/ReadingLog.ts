import { normalize } from 'normalizr';
import * as schema from './../schema';
import { call, put } from 'redux-saga/effects';
import { ActionTypes, ReviewStatus } from '../interfaces';
import * as apis from '../apis';
import * as actions from '../actions';

export function* fetch({ payload: { limit } }: actions.ReadingLog.FetchAll) {
  try {
    const data = yield call(apis.fetch, `/api/readingLogs`, {
      params: { limit, _expand: ['book', 'review'] },
    });
    const { entities, result } = normalize(data, [schema.readingLog]);
    yield put({
      type: ActionTypes.FETCH_SUCCESS,
      payload: { entities, ui: { readingLogs: result } },
    });
  } catch (error) {
    yield put({ type: ActionTypes.FETCH_FAILURE, payload: error });
  }
}

export function* create({ payload: { title } }: actions.ReadingLog.Create) {
  try {
    const bookData = yield call(apis.getBookData, title);
    const book = yield call(apis.post, '/api/books', bookData);
    const review = yield call(apis.post, '/api/reviews', {
      status: ReviewStatus.PLAN_TO_BUY,
      rating: 0,
    });

    const data = yield call(apis.post, '/api/readingLogs', {
      id: book.id,
      bookId: book.id,
      reviewId: review.id,
    });
    data.book = book;
    data.review = review;

    const { entities, result } = normalize(data, schema.readingLog);
    yield put({
      type: ActionTypes.FETCH_SUCCESS,
      payload: { entities, ui: { readingLogs: [result] } },
    });
  } catch (error) {
    yield put({ type: ActionTypes.FETCH_FAILURE, payload: error });
  }
}

export function* remove({
  payload: { readingLogId },
}: actions.ReadingLog.Remove) {
  try {
    yield call(apis.remove, `/api/readingLogs/${readingLogId}`);
    yield put({
      type: ActionTypes.REMOVE_READING_LOG_SUCCESS,
      payload: { readingLogId },
    });
  } catch (error) {
    yield put({ type: ActionTypes.FETCH_FAILURE, payload: error });
  }
}
