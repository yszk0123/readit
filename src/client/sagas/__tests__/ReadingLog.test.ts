import { call } from 'redux-saga/effects';
import { ReviewStatus } from '../../interfaces';
import { post, getBookData } from '../../apis';
import * as Actions from '../../actions';
import * as ReadingLog from '../ReadingLog';

describe('ReadingLog', () => {
  it('should create', () => {
    const title = 'book-title';
    const gen = ReadingLog.create(Actions.ReadingLog.create({ title }));
    expect(gen.next().value).toEqual(call(getBookData, title));

    const bookData = { id: 'book-1' };
    expect(gen.next(bookData).value).toEqual(
      call(post, '/api/books', bookData),
    );

    const book = { id: 'book-1' };
    const reviewData = { status: ReviewStatus.PLAN_TO_BUY, rating: 0 };
    expect(gen.next(book).value).toEqual(
      call(post, '/api/reviews', reviewData),
    );

    const review = { id: 'review-1' };
    const readingLogData = {
      id: book.id,
      bookId: book.id,
      reviewId: review.id,
    };
    expect(gen.next(review).value).toEqual(
      call(post, '/api/readingLogs', readingLogData),
    );

    const data = { id: 'readingLog-1' };
    expect(gen.next(data).value).toMatchSnapshot();
  });
});
