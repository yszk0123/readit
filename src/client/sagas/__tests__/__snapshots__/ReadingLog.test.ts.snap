// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadingLog should create 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "entities": Object {
          "books": Object {
            "book-1": Object {
              "id": "book-1",
            },
          },
          "readingLogs": Object {
            "readingLog-1": Object {
              "book": "book-1",
              "id": "readingLog-1",
              "review": "review-1",
            },
          },
          "reviews": Object {
            "review-1": Object {
              "id": "review-1",
            },
          },
        },
        "ui": Object {
          "readingLogs": Array [
            "readingLog-1",
          ],
        },
      },
      "type": "FETCH_SUCCESS",
    },
    "channel": null,
  },
}
`;
